nvdaHelper/remote/log.cpp

```diff
@@ -14,25 +14,39 @@ This license can be found at:
 
 #include <string>
 #include <queue>
+#include <mutex>
 #include <crtdbg.h>
 #include "nvdaControllerInternal.h"
 #include "nvdaHelperRemote.h"
 #include <common/log.h>
 
 std::deque<std::tuple<int, std::wstring>> logQueue;
+std::mutex logQueueLock;
 
-// Fetch a  maximum number of messages from the log queue
+// Fetch all available messages from the queue
 // and send them onto NvDA via rpc.
-void __stdcall log_flushQueue(ULONG_PTR maxCount) {
-	while(!logQueue.empty()) {
-		if(maxCount < 1) break;
-		{
-			auto& [level, msg] = logQueue.front();
-			nvdaControllerInternal_logMessage(level, GetCurrentProcessId(), msg.c_str());
-		}
-		logQueue.pop_front();
-		--maxCount;
+void log_flushQueue() {
+	// Ensure this is never called from outside the manager thread.
+	if(
+		!inprocMgrThreadHandle
+		|| GetCurrentThreadId() != GetThreadId(inprocMgrThreadHandle)
+	) {
+		return;
+	}
+	std::deque<std::tuple<int, std::wstring>> tempQueue;
+	{
+		std::lock_guard lock{logQueueLock};
+		tempQueue.swap(logQueue);
 	}
+	while(!tempQueue.empty()) {
+		auto& [level, msg] = tempQueue.front();
+		nvdaControllerInternal_logMessage(level, GetCurrentProcessId(), msg.c_str());
+		tempQueue.pop_front();
+	}

```

Comment:
Due to the introduction of the swap with the tempQueue, this no longer needs to remove elements one by one. It can allow the whole tempQueue to be destroyed at once.

```suggestion
	for(auto& [level, msg] : tempQueue) {
		nvdaControllerInternal_logMessage(level, GetCurrentProcessId(), msg.c_str());
	}

```